Cdm overhaul

asinghvi17 opened the pull request.

rafaqz committed:
Fix outbounds line burn (#951)

* fix out of bounds line burning

* no show

* omg

rafaqz committed:
bump patch version to 0.14.4

tiemvanderdeure committed:
use the correct `@example` label

tiemvanderdeure committed:
Merge pull request #953 from tiemvanderdeure/bonus_spatialmean

Fix spatial mean bonus section

rafaqz committed:
start cdm read overhaul

asinghvi17 committed:
Fix multilinestring load

asinghvi17 committed:
Fix everything for CDM geom lookup + add tests

rafaqz committed:
more cf read / write

asinghvi17 committed:
namespace coord names in geometry lookup, but only if necessary

rafaqz committed:
read geom working

rafaqz committed:
round trip geometrylookup

rafaqz committed:
working Raster

rafaqz committed:
add cf test files

asinghvi17 commented:
@rafaqz is this ready to merge into the geometry lookup branch from your end?

asinghvi17 commented:
Yea that works too

rafaqz committed:
add more cf examples

rafaqz commented:
35 CF examples now! - all loading but not all in the optimal way. Rotations are still missing, and I'm not sure how to handle these lookups with a center point and an explicit polygon boundary.

With a regular lookup `At` would select the center point, and `Contains` would select anything in the cell (we should have used `Coveredby` for that, need to DE-I9M-ize DD asap)

So for this it could do the same thing? and maybe for all geometry lookups? `At` gets the centroid/representative point, and `Contains` or whatever checks if the point is in the area.

(I'm thinking we might need a DE19M.jl or DE19MPredicates.jl package to hold all of these wrapper types, then we can use them everywhere, like `Extents.predicate(Contains(), a, b)` and we have the predicate separated from the context.)

asinghvi17 commented:
> At gets the centroid/representative point, and Contains or whatever checks if the point is in the area.

I don't know about that....it seems pretty useless, the representative point must be within the geometry at any rate

asinghvi17 commented:
> DE19MPredicates.jl

That would be nice...but would conflict with GeometryOps operations when we add those.  Plus Rasters has Touches which is not DE9IM touches...

rafaqz commented:
Yeah swapping DD touches and covers is where the idea can from. I don't think it will conflict, are there types in GO? We can just use them there too?

asinghvi17 commented:
> Haha then we can select with geometries omg

We already can ;)

> are there types in GO

No but there will be.  And they have to subtype `GOCore.Operation`.  That's so that they fit in the whole declarative API, if not then it's not really feasible...

One thing we can do is that DE9IMPreds.jl depends on GOCore, and Extents, and actually fills in the Extents.predicate method for these predicates.  But then DD depends on GOCore, not sure if we want that.

rafaqz commented:
> it seems pretty useless

Maybe, but I'm not sure what else `At` would do, probably error.

rafaqz commented:
> We already can ;)

But like, with all the predicates even on a regular raster

asinghvi17 commented:
The problem there is again the names will clash

asinghvi17 commented:
> But like, with all the predicates even on a regular raster

what does that mean?  If I say Intersects(geom), then it can only give me either the bbox or a mergedlookup raster, neither of which are particularly nice.

How do we specify that the raster should be masked?  Maybe a new MaskingIndexSet type?

rafaqz commented:
I don't get why we need both objects, the idea was to use the same thing rather than clash...

asinghvi17 commented:
I think we're thinking of two different objects.  I'm thinking of the GO Operations interface, i.e.

```julia
abstract type GeometricPredicate <: Operation end

struct Contains{Alg{M}, Exact, ...} <: GeometricPredicate end

nargs(::Contains) = 2
application_level(::Contains) = GO.TraitTarget{GI.AbstractGeometryTrait}()

function (::Contains)(::GI.PointTrait, p1, ::GI.PointTrait, p2)
end
```

or something like this

***

My understanding is that you are thinking of more of a DD Dimension or Selector like struct?

The issue that I have is that `GO.Contains() != DE9IM.Contains()`, which will inevitably lead to name clashes basically everywhere.  So we should figure out how to mitigate that.  Maybe we just say that predicates are not GO operations?  Not sure what the best solution is here.  But this is the issue in my mind.

I don't know how the Operation interface will turn out - maybe we can make it semi trait based.  Or maybe we don't end up doing it at all?  It's kind of tricky.

rafaqz commented:
Were not ;)

We can put all those details in another object that goes inside the predicate wrapper.

Then they can be shared. 

DD uses Dimension wrappers for a lot of unrelated things in exactly this way. You can even pass keywords to a Dimension and they can get found and constructed later depending on the context, e.g. `X(data; sampling=Intervals())` works.

asinghvi17 committed:
Update make.jl

asinghvi17 committed:
add my DV branch for now

asinghvi17 committed:
REMOVE BEFORE MERGING: draft=true for faster iteration

asinghvi17 committed:
Update make.jl

asinghvi17 committed:
Update docs/Project.toml

rafaqz committed:
more cf

tiemvanderdeure committed:
warnonly=false and import packages (#954)

alex-s-gardner committed:
move logo to top right

alex-s-gardner committed:
add new aqua logo

alex-s-gardner committed:
remove broken downloads

alex-s-gardner committed:
modify text

alex-s-gardner committed:
fix Downloads badge

alex-s-gardner committed:
fix downloads badge request

lazarusA committed:
Merge pull request #977 from alex-s-gardner/update-ReadMe

 Beautify ReadMe.. a bit

asinghvi17 committed:
Update docs/make.jl

asinghvi17 committed:
Update config.mts

asinghvi17 committed:
Merge branch 'main' into asinghvi17-patch-7

lazarusA committed:
Merge pull request #974 from rafaqz/asinghvi17-patch-7

rafaqz committed:
fix lazy in create without filename (#979)

* fix lazy in create without filename

* fix Raster ambiguity

rafaqz committed:
bump patch version to 0.14.5

tiemvanderdeure committed:
remove unnecessary `replace_missing` (#982)

rafaqz committed:
check if dimtree exists (#985)

rafaqz committed:
bump patch version to 0.14.6

rafaqz committed:
Merge branch 'main' into cdm_overhaul

tiemvanderdeure committed:
fix grib missingval (#966)

tiemvanderdeure committed:
test aggregate without allowscalar (#990)

rafaqz committed:
more passing cf

rafaqz committed:
add UGRID mesh

tiemvanderdeure committed:
fix OpenStack calls (#968)

* fix OpenStack calls

* add a test

* fix test

rafaqz committed:
refactor cf

rafaqz committed:
mostly working climatology bounds

rafaqz committed:
unaligned

rafaqz committed:
better rotated lookups

asinghvi17 committed:
Update Makie and GB compat

asinghvi17 committed:
bump patch version

asinghvi17 committed:
Merge pull request #992 from rafaqz/asinghvi17-patch-8

Update Makie and GB compat

asinghvi17 committed:
Merge pull request #993 from rafaqz/asinghvi17-patch-9

bump patch version

asinghvi17 committed:
correct patch version for registration (#994)

asinghvi17 committed:
Merge remote-tracking branch 'origin/main' into cdm_overhaul (#996)

asinghvi17 committed:
fix sources for dev branches

tiemvanderdeure committed:
drop some if-else blocks for julia < 1.9 (#998)

* drop some if-else blocks for julia < 1.9

* revert changes to runtests.jl (oops)

RuiRojo committed:
Fix missing return arg in edge burning (#1007)

Co-authored-by: Rui Rojo <rui.rojo@proton.me>

dependabot[bot] committed:
Bump actions/checkout from 4 to 5 (#1005)

Bumps [actions/checkout](https://github.com/actions/checkout) from 4 to 5.
- [Release notes](https://github.com/actions/checkout/releases)
- [Changelog](https://github.com/actions/checkout/blob/main/CHANGELOG.md)
- [Commits](https://github.com/actions/checkout/compare/v4...v5)

---
updated-dependencies:
- dependency-name: actions/checkout
  dependency-version: '5'
  dependency-type: direct:production
  update-type: version-update:semver-major
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

rafaqz committed:
fix #999 (#1003)

Co-authored-by: Raf Schouten <schoutenr@ull-pf39vwmc.mobility.unimelb.net.au>

rafaqz committed:
fix empty RasterStack (#1004)

* fix empty RasterStack

* fix error test

---------

Co-authored-by: Raf Schouten <schoutenr@ull-pf39vwmc.mobility.unimelb.net.au>

rafaqz committed:
Make JET happier (#961)

* bugfix with JET

* jet things

* fix mosaic

* fix GDAL tif driver

* fix feature rasterize

* remove stray show

rafaqz committed:
domains working

rafaqz committed:
fix lazy Char string reading and other minor things

alex-s-gardner committed:
Update README.md (#1010)

* Update README.md

- move doc badges above other badges for improved discovery 
- improve intro and key features text

* Update README.md

Co-authored-by: Anshul Singhvi <anshulsinghvi@gmail.com>

* Update README.md

Co-authored-by: Anshul Singhvi <anshulsinghvi@gmail.com>

---------

Co-authored-by: Anshul Singhvi <anshulsinghvi@gmail.com>

rafaqz committed:
start testing write loop

asinghvi17 committed:
Make sure `open` is always called in aggregate (#1021)

* Make sure `open` is always called in aggregate

This makes it such that the `src` is first opened and then cached.  Also changed 2 variable names to enhance readability.

Someone should probably add a test.  But at least locally this fixes #1020 for me.

* add additional lazy aggregate test (#1022)

---------

Co-authored-by: Alex Gardner <32276930+alex-s-gardner@users.noreply.github.com>

tiemvanderdeure committed:
fix coverage! and improve performance (#1013)

* fix and test coverage!

* improve performance of coverage by taking a view

* return original raster instead of view

* use archive unidata

* use archive unidata in docs

* fix _sum_coverage threaded

Co-authored-by: Alex Gardner <alex.s.gardner@jpl.nasa.gov>

* fix rasterizing to dimtuple

* verbose not vebose

* test rasterizing to dims

* fix coverage for 1 geometry

* test coverage on single geometry

* fix coverage on a single geometry

---------

Co-authored-by: Alex Gardner <alex.s.gardner@jpl.nasa.gov>

tiemvanderdeure committed:
fix threadid for 1.12 (#1030)

* fix threadid for 1.12

* rewrite to use channels instead of threadid

* git ignore all manifest files

* fix

* use a _maybe_channel function and get rid of :static

* cleaner code in extract

* fix threaded coverage for nthreads = 1

* fix convertedlookup tests

* fix gdal tests

* try to fix docs

* use GI.isempty instead of GI.npoint

* fix eltype of channel

* add a function barrier in union coverage

* revert changes to plotting

* specialize do_broadcast for a single argument

* specialize with_resource

Co-authored-by: Anshul Singhvi <anshulsinghvi@gmail.com>

* lazy iteration in sum

Co-authored-by: Anshul Singhvi <anshulsinghvi@gmail.com>

---------

Co-authored-by: Anshul Singhvi <anshulsinghvi@gmail.com>

rafaqz committed:
bump patch version to 0.14.7

asinghvi17 committed:
Actually support user-defined `op` in `rasterize` (#1033)

* Actually respect user's preference of `op`

How have we not come across this before?

* Only get `zero(filleltype)` if eltype is not known

* if statement -> ternary

Co-authored-by: Tiem van der Deure <tiemvanderdeure@gmail.com>

* Add a test for the `vcat` op

With a vector of Ints and a specified eltype.  Just to make sure this pipeline works.

* Add LTS CI

* use modern CI header from template

* Cheeky crazy rasterization tutorial

* Delete bad theme reference

We need to clean up that theme, apparently

* fix plotting tutorial by namespacing

* add a comment about what the fill expression is doing

Co-authored-by: Rafael Schouten <rafaelschouten@gmail.com>

* Update docs/src/tutorials/crazy_rasterization.md

Co-authored-by: Rafael Schouten <rafaelschouten@gmail.com>

* Update docs/src/tutorials/crazy_rasterization.md

Co-authored-by: Rafael Schouten <rafaelschouten@gmail.com>

* Update docs/src/tutorials/crazy_rasterization.md

* More plotting tutorial fixes

* Fix Documenter concurrency group in CI

* Deploy docs even if there are errors

* print correctly

* enclose bad video in comments

* Don't fail fast in CI

* CI on 1.11 too

* Merge branch 'main' into as/flexible-rasterize

* Fix 1.11 CI by checking only for AbstractMatrix-ness

I guess 1.12 is able to drop the wrapper and the extra dim but 1.11 is not

* Fix Rasters.sample on LTS (#1035)

Co-authored-by: Anshul Singhvi <anshulsinghvi@gmail.com>

---------

Co-authored-by: Tiem van der Deure <tiemvanderdeure@gmail.com>
Co-authored-by: Rafael Schouten <rafaelschouten@gmail.com>

tiemvanderdeure committed:
Do not use `Base.get_extension` (#1027)

* add Extension abstract type and is_loaded function

* swap if logic to drop a !

* Use singleton instances instead of types

---------

Co-authored-by: Anshul Singhvi <anshulsinghvi@gmail.com>
Co-authored-by: Rafael Schouten <rafaelschouten@gmail.com>

alex-s-gardner committed:
Update CommonDataModel compat version to 0.4 (#1036)

alex-s-gardner committed:
Add CommonDataModel version 0.3 back into compat (#1038)

tiemvanderdeure committed:
make sure threadsafe rasterization ops are efficient (#1039)

* add and test commutative ops

* CommutativeOps

* use test_logs

* also test count and mean

* drop a line outside testset

* Don't disable logging, see what happens

---------

Co-authored-by: Anshul Singhvi <anshulsinghvi@gmail.com>

tiemvanderdeure committed:
more BackendException less error() (#1043)

tiemvanderdeure committed:
simplify sourceconstructor dispatch (#1044)

tiemvanderdeure committed:
Fix finding files for RasterSeries with a pattern (#1024)

* use startswith instead of indexing

* test finding filenames with a pattern

* sort files before testing equality

* filter_ext is not base

tiemvanderdeure committed:
Fix flipped outputs from resample (#1046)

* reorder before assigning dims in

* test resample with different dimension order

* add missing `@test`

* fix docs logging

felixcremer committed:
Add force to RasterCreator (#1048)

* Add force keyword to RasterCreator

* Make temporary_random_rasters use a temporary dirname

This is so that other tif files in the tempdir is not interfering with the Raster tests.

github-actions[bot] committed:
CompatHelper: bump compat for CFTime in [extras] to 0.2, (keep existing compat) (#1052)

Co-authored-by: CompatHelper Julia <compathelper_noreply@julialang.org>

github-actions[bot] committed:
CompatHelper: bump compat for Shapefile in [extras] to 0.13, (keep existing compat) (#1051)

Co-authored-by: CompatHelper Julia <compathelper_noreply@julialang.org>

dependabot[bot] committed:
Bump actions/checkout from 5 to 6 (#1049)

Bumps [actions/checkout](https://github.com/actions/checkout) from 5 to 6.
- [Release notes](https://github.com/actions/checkout/releases)
- [Changelog](https://github.com/actions/checkout/blob/main/CHANGELOG.md)
- [Commits](https://github.com/actions/checkout/compare/v5...v6)

---
updated-dependencies:
- dependency-name: actions/checkout
  dependency-version: '6'
  dependency-type: direct:production
  update-type: version-update:semver-major
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

github-actions[bot] committed:
CompatHelper: bump compat for GeoDataFrames in [extras] to 0.4, (keep existing compat) (#1053)

Co-authored-by: CompatHelper Julia <compathelper_noreply@julialang.org>
Co-authored-by: Rafael Schouten <rafaelschouten@gmail.com>

github-actions[bot] committed:
CompatHelper: bump compat for CommonDataModel to 0.4, (keep existing compat) (#1015)

Co-authored-by: CompatHelper Julia <compathelper_noreply@julialang.org>
Co-authored-by: Rafael Schouten <rafaelschouten@gmail.com>

rafaqz committed:
add `emptyval` keyword to `zonal` (#1023)

* zonal_emptyval

* remove accidental coverage changes

rafaqz committed:
force floating point for `aggregate` `mean` (#1028)

* force floating point for aggregate mean

* also use add_sum and mul_prod to define init conditions

* Update src/methods/aggregate.jl

Co-authored-by: Tiem van der Deure <tiemvanderdeure@gmail.com>

* add tests

* use Base.agg_sum in all agg

* bugfix add_sum

* bugfix rasterize

* bugfix

---------

Co-authored-by: Tiem van der Deure <tiemvanderdeure@gmail.com>
Co-authored-by: Anshul Singhvi <anshulsinghvi@gmail.com>

rafaqz committed:
update DD to 0.30 and bugfix (#1059)

* update DD to 0.30 and bugfix

* bugfix

* add order

---------

Co-authored-by: Tiem van der Deure <tiemvanderdeure@gmail.com>

rafaqz committed:
Merge branch 'breaking' into cdm_overhaul

rafaqz committed:
working crs

rafaqz committed:
more cdm

dependabot[bot] committed:
Bump julia-actions/setup-julia from 2 to 3 (#1069)

Bumps [julia-actions/setup-julia](https://github.com/julia-actions/setup-julia) from 2 to 3.
- [Release notes](https://github.com/julia-actions/setup-julia/releases)
- [Commits](https://github.com/julia-actions/setup-julia/compare/v2...v3)

---
updated-dependencies:
- dependency-name: julia-actions/setup-julia
  dependency-version: '3'
  dependency-type: direct:production
  update-type: version-update:semver-major
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
Bump codecov/codecov-action from 5 to 6 (#1068)

Bumps [codecov/codecov-action](https://github.com/codecov/codecov-action) from 5 to 6.
- [Release notes](https://github.com/codecov/codecov-action/releases)
- [Changelog](https://github.com/codecov/codecov-action/blob/main/CHANGELOG.md)
- [Commits](https://github.com/codecov/codecov-action/compare/v5...v6)

---
updated-dependencies:
- dependency-name: codecov/codecov-action
  dependency-version: '6'
  dependency-type: direct:production
  update-type: version-update:semver-major
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

tiemvanderdeure committed:
Breaking: rename projectedindex and mappedindex to projectedlookup and mappedlookup (#1029)

* replace index with lookup a bunch of places

* Update test/array.jl

Co-authored-by: Rafael Schouten <rafaelschouten@gmail.com>

* Update src/lookup.jl

Co-authored-by: Rafael Schouten <rafaelschouten@gmail.com>

* Update src/lookup.jl

Co-authored-by: Rafael Schouten <rafaelschouten@gmail.com>

* revert a silly change to ncdatasets tests

* parent lookup 1

Co-authored-by: Rafael Schouten <rafaelschouten@gmail.com>

* parent lookup 2

Co-authored-by: Rafael Schouten <rafaelschouten@gmail.com>

* parent lookup in plotrecipes

* fix a bug

* fix plotrecipes

* projectedindex -> projectedlookup

---------

Co-authored-by: Rafael Schouten <rafaelschouten@gmail.com>

tiemvanderdeure committed:
improve disaggregate performance (#1017)

* use fill! instead of .= for type stability

* fix disaggregate to file test

---------

Co-authored-by: Anshul Singhvi <anshulsinghvi@gmail.com>
Co-authored-by: Rafael Schouten <rafaelschouten@gmail.com>

dependabot[bot] committed:
Bump julia-actions/cache from 2 to 3 (#1066)

Bumps [julia-actions/cache](https://github.com/julia-actions/cache) from 2 to 3.
- [Release notes](https://github.com/julia-actions/cache/releases)
- [Commits](https://github.com/julia-actions/cache/compare/v2...v3)

---
updated-dependencies:
- dependency-name: julia-actions/cache
  dependency-version: '3'
  dependency-type: direct:production
  update-type: version-update:semver-major
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

github-actions[bot] committed:
CompatHelper: bump compat for GRIBDatasets in [extras] to 0.5, (keep existing compat) (#1072)

Co-authored-by: CompatHelper Julia <compathelper_noreply@julialang.org>

rafaqz committed:
Merge branch 'main' into breaking

rafaqz committed:
add SRTM extension for RDS (#1078)

rafaqz committed:
add SRTM extension for RDS (#1078)

rafaqz committed:
add CDM 0.5

rafaqz committed:
comment out broken CFTime example

rafaqz committed:
use setdims

tiemvanderdeure committed:
Breaking: more consistent (dis)aggregation of lookups (#914)

* shift locus before disaggregating a lookup

* add test

* more proper fix

* test that extent is preserved

* remove stray show

Co-authored-by: Felix Cremer <felix.cremer@dlr.de>

* tweak tests

* add a comment to `ceil`

* handle aggregation of irregular lookups

* fix a method ambiguity

* fix another method ambiguity

* index => lookup

---------

Co-authored-by: Felix Cremer <felix.cremer@dlr.de>
Co-authored-by: Rafael Schouten <rafaelschouten@gmail.com>

rafaqz committed:
Stable ranges (#1075)

* Fixed range index bug

* fix range type

* Apply suggestion from @rafaqz

* Apply suggestion from @rafaqz

* anchored range

* use anchored range in aggregate too

* comment

* test against GDAL

* test against GDAL with 1 ups tolerance

* Apply suggestion from @rafaqz

* Apply suggestion from @rafaqz

* use StableRange

* add view/dotview methods

* more bugfix

* remove redundant TwicePrecision wrapper

Co-authored-by: Tiem van der Deure <tiemvanderdeure@gmail.com>

* use StableRange more widely, and allow stop inputs like StepRangeLen

* broadcast methods for StableRange

---------

Co-authored-by: dnstudent <davide.nicoli@unimi.it>
Co-authored-by: Tiem van der Deure <tiemvanderdeure@gmail.com>

felixcremer committed:
[Breaking] Set missingval of count rasterize to non-zero after rasterisation (#1067)

* update DD to 0.30 and bugfix

* bugfix

* Set rasterize count missingval to notzero

* Use MockChunkedDiskArray instead of RechunkedDiskArray in test

* Fix tests for rasterization

---------

Co-authored-by: Rafael Schouten <rafaelschouten@gmail.com>

rafaqz committed:
Merge pull request #1076 from rafaqz/breaking

Breaking

rafaqz committed:
all passing

rafaqz committed:
Merge branch 'main' into cdm_overhaul

rafaqz committed:
reorganise and test round-tripping more comprehensively
